### ISSUE TITLE
emacs hangs after evil-open-below in org-mode

### ISSUE THREAD

**Issue by dwuggh** [syl20bnr/spacemacs#13684]
#### Description :octocat:
emacs hangs after evil-open-below in org-mode, I think it could be a bug in the line number counting process

#### Reproduction guide :beetle:
- Start Emacs
- create a .org file
- write something like
  ```org
  * foo
    - bar
      1. put your cursor at this line
          barfoo
      2. foofoo
  ```
- place the cursor at line "put your cursor at this line"
- press "o"(evil-open-below)

*Observed behaviour:* :eyes: :broken_heart:
emacs hanged for about 5 mins(and then I killed it)

*Expected behaviour:* :heart: :smile:
this should not happen

#### System Info :computer:
- OS: gnu/linux
- Emacs: 27.0.91
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. ecba7ad0d)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: ivy
- Layers:
```elisp
(csv yaml vimscript emacs-lisp git ivy
     (auto-completion :variables auto-completion-enable-snippets-in-popup t)
     lsp markdown multiple-cursors
     (org :variables org-agenda-files
          '("~/org/agenda")
          org-enable-org-journal-support t org-journal-dir "~/org/journal/" org-journal-enable-agenda-integration t)
     (mylatex :variables latex-enable-magic t latex-enable-folding t latex-build-command "LatexMk")
     pdf orgwithlatex
     (spell-checking :variables ispell-program-name "aspell" spell-checking-enable-by-default nil)
     syntax-checking treemacs
     (c-c++ :variables c-c++-backend 'lsp-ccls)
     (cmake :variables cmake-enable-cmake-ide-support t)
     rust
     (python :variables python-backend 'lsp python-lsp-server 'mspyls)
     prettier
     (javascript :variables javascript-backend 'lsp javascript-lsp-linter nil javascript-fmt-tool 'prettier)
     (typescript :variables typescript-fmt-tool 'prettier typescript-linter 'eslint typescript-backend 'tide typescript-lsp-linter nil)
     (vue :variables vue-backend 'lsp)
     html semantic dap themes-megapack dwuggh language
     (haskell :variables haskell-completion-backend 'lsp))
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GLIB NOTIFY INOTIFY ACL GNUTLS LIBXML2 FREETYPE HARFBUZZ LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD JSON PDUMPER LCMS2 GMP

#### Backtrace :paw_prints:
```
```

**Comment by cmatzenbach** [comment]
Yeah I can confirm this as well.  Similar System Info (differences below), same example, same behavior.  It really crashes emacs for me, it's been about 5 minutes and I'm on a black, single-color screen with no buffer anymore.

Differences in System Info:
* Emacs 26.3
* Spacemacs branch: develop (rev 2bc80b4b4)

Layers obviously differs, but I don't think it has any affect on this bug - seems to be some collision between `(evil-open-below)` and org layer.

Fwiw, on my slightly older version of emacs, this has caused it to hang forever and not respond to normal OS exit functions.  Going to have to kill the process directly.  Definitely seems like a high priority for the commonly-used org-layer.

**Comment by cmatzenbach** [comment]
Can someone with proper access set this to high priority?  I don't think we need much further testing - the bug is definitely present.

**Comment by dwuggh** [comment]
This could be a bug in evil-org. I have a temporary workaround:
```elisp
(advice-add 'evil-org-open-below :override
            (lambda (count)
              (interactive "P")
              (evil-next-line)
              (evil-org-open-above count)))
```

**Comment by emacs18** [comment]
Following is the stack trace I get on emacs 27. The code seems to be stuck in infinite loop within nested while loops within org-list-struct-apply-struct function.

Perhaps this is a bug in org-mode.

```
Debugger entered--Lisp error: (quit)
  (while (< (point) down) (if (looking-at-p "[ \11]*$") nil (setq min-ind (min (current-indentation) min-ind)) ...
  (save-excursion (goto-char up) (while (< (point) down) (if (looking-at-p "[ \11]*$") nil (setq min-ind ...
  (let* ((child (cdr (assq up acc-end))) (ind (org-list-get-ind child struct)) (min-ind most-positive-fixnum)) ...
  (if itemp (cdr (assq up itm-shift)) (let* ((child (cdr (assq up acc-end))) (ind (org-list-get-ind child struct)) ...
  (let* ((up (car-safe (prog1 all-ends (setq all-ends (cdr all-ends))))) (down (car all-ends)) (itemp (assq up struct)) ...
  (while (cdr all-ends) (let* ((up (car-safe (prog1 all-ends (setq all-ends (cdr all-ends))))) (down (car all-ends)) ...
  (let (end-list acc-end itm-shift all-ends sliced-struct) (let ((--dolist-tail-- old-struct)) ...
  (let* ((origin (point-marker)) (inlinetask-re (and (featurep 'org-inlinetask) ...
  org-list-struct-apply-struct(((7 2 "- " nil nil nil 88) (15 4 "1. " nil nil nil 51) (51 4 "2. " nil nil nil 74) ...
  (let ((old-struct (or old-struct (copy-tree struct)))) (org-list-struct-fix-ind struct parents 2) ...
  org-list-write-struct(((7 2 "- " nil nil nil 88) (15 4 "1. " nil nil nil 51) (51 4 "2. " nil nil nil 74) ...
  (let* ((struct (save-excursion (goto-char itemp) (org-list-struct))) (prevs (org-list-prevs-alist struct)) ...
  (if (save-excursion (goto-char itemp) (org-at-item-timer-p)) (progn (org-timer-item) t) (let* ((struct (save-excursion ...
  (if (or (not itemp) (save-excursion (goto-char itemp) (org-invisible-p))) nil (if (save-excursion (goto-char itemp) ...
  (let ((itemp (org-in-item-p)) (pos (point))) (if (or (not itemp) (save-excursion (goto-char itemp) (org-invisible-p))) ...
  org-insert-item(nil)
  (progn (end-of-visible-line) (org-insert-item (org-at-item-checkbox-p)))
  (and (memq 'item evil-org-special-o/O) (org-at-item-p) (progn (end-of-visible-line) (org-insert-item (org-at-item-checkbox-p))))
  (cond ((and (memq 'table-row evil-org-special-o/O) (org-at-table-p)) (org-table-insert-row '(4)) (evil-insert nil)) ...
  evil-org-open-below(nil)
  funcall-interactively(evil-org-open-below nil)
  call-interactively(evil-org-open-below nil nil)
  command-execute(evil-org-open-below)
```

**Comment by duianto** [comment]
Confirmed.

After pressing: `o`
Spacemacs hangs and the cpu usage of the Emacs process goes up:
- in PopOS, to: `50%`
- in Windows, to: `25-27%` and the Power usage changes to: `Very High`

It stops when `C-g` is pressed.
But now a second `1.` list item has been added, and the cursor is in evil normal state, in the left most column.
```
* foo
  - bar
    1. put your cursor at this line
    1. 
        barfoo
    2. foofoo
```

### It doesn't seem to happen without Spacemacs

With just Emacs and the packages: `evil` and `evil-org`

Pressing: `o`
Adds a new list item:
- with the correct list numbering,
- the cursor is in evil insert state, below the p in `put` (as expected)
```elisp
* foo
  - bar
    1. put your cursor at this line
    2. 
        barfoo
    3. foofoo
```

<details><summary>
PopOS 20.04, System Info (click to expand)
</summary>
<pre>
#### System Info :computer:
- OS: gnu/linux
- Emacs: 26.3.50
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. 6145a566e)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(auto-completion emacs-lisp git helm multiple-cursors org python
                 (shell :variables shell-default-shell 'eshell shell-default-height 30 shell-default-position 'bottom)
                 version-control treemacs)
```
- System configuration features: XPM JPEG TIFF GIF PNG SOUND DBUS GSETTINGS GLIB NOTIFY LIBSELINUX GNUTLS FREETYPE XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS

</pre>
</details>

<details><summary>
Windows 1903, System Info (click to expand)
</summary>
<pre>
#### System Info :computer:
- OS: windows-nt
- Emacs: 26.3
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. d4634b571)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(autohotkey auto-completion emacs-lisp git helm html javascript
            (latex :variables latex-build-command "LatexMk" latex-enable-magic t)
            (markdown :variables markdown-live-preview-engine 'vmd markdown-command "vmd")
            multiple-cursors
            (org :variables org-agenda-files
                 '("~/org/notes.org"))
            python ruby
            (shell :variables shell-default-shell 'eshell shell-default-height 30)
            spell-checking
            (syntax-checking :variables syntax-checking-enable-by-default nil)
            themes-megapack treemacs version-control)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG SOUND NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS THREADS LCMS2

</pre>
</details>

**Comment by duianto** [comment]
It seems to be caused by the command: `(org-insert-item)`
when the text below the list item: `1. put your cursor at this line`
is indented to start below the period or more:

### No issue
```
    1. put your cursor at this line
a
```
```
    1. put your cursor at this line
    a
```
or anywhere in between.

### Issue
```
    1. put your cursor at this line
     a
```
or if it's indented more.

The issue occurs with the indentation above, when the cursor is on either the `1. put your ...` line,
or on the indented line below.
And the insert item command is evaluated: `S-M-;` `(org-insert-item)`

**Comment by duianto** [comment]
It doesn't seem to be `evil` related, it happens with the `emacs` editing style and/or when `evil-mode` is disabled.

And it also happens with the `spacemacs-base` distribution where a lot of packages are disabled.

**Comment by duianto** [comment]
It doesn't seem to be caused by Spacemacs.

Because it happens with just Emacs and the `org-plus-contrib` package.

@bzg Is this a known issue?
Scroll down to the `Issue` section below, since you probably have Emacs and `org-plus-contrib` setup already 😄 

### Setup
Backup the current Spacemacs setup,
by renaming or moving the following directory and files somewhere else:
- Directory: `~/.emacs.d`
- Files: `~/.spacemacs` and `~/.spacemacs.env`

#### Installing org-plus-contrib
- Copy these lines:
```
(progn
  (require 'package)
  (add-to-list 'package-archives '("org" . "http://orgmode.org/elpa/") t))
```
source: https://emacs.stackexchange.com/a/29116
- Start Emacs
- Evaluate the copied expression:
  - Open the Eval prompt: `S-M-;`
  - Paste and evaluate the copied lines: `C-y` `RET`
- Open the packages buffer: `M-x packages-list-packages RET`
The packages list updates and the mode line shows: `Menu:Loading`
wait until when the minibuffer (at the bottom) shows the message: `Package refresh done`
- Open the search prompt: `C-s`
- Type: `org-plus`
- Highlight the line with the `org-plus-contrib` package by pressing: `C-s`
- Press: `RET` to close the search prompt and leave the cursor on the `org-plus-contrib` line
- Mark it for installation by pressing: `i`
- Install the package by pressing: `x`
- Confirm the package installation by pressing: `y`

The latest version of org is now installed.

### Issue
This might work as a minimal example:
- Copy the following lines:
```
1. a
   b
2. c
```
- Open a new buffer: `C-x b` `test.org` `RET`
- Switch to org mode: `M-x` `org-mode` `RET`
- Paste the copied lines: `C-y`
- Move the cursor to the end of the first line
- Evaluate the following command: `S-M-;` `(org-insert-item)` `RET`
The only visible change, is that the cursor moves to the beginning of the `Eval:` prompt in the minibuffer, 
The Emacs cpu usage spikes.
- The excess cpu usage can stopped by pressing: `C-g`
The buffer shows:
```
1. a
1. 
   b
2. c
```

The example above has three spaces before the second lines: `b`
The same soft freeze (meaning that it can be canceled with `C-g`) is observed if more spaces are added before the `b`.

#### `b` indented with two or three spaces

Emacs shows an error instead of freezing, if the second lines indentation is reduced to two or three spaces.

Now when evaluating: `(org-insert-item)` from the end of the first line,
Then a incorrectly numbered list item is added:
```
1. a
1. 
 b
2. c
```
And a Backtrace buffer opens with the message:
```
Debugger entered--Lisp error: (args-out-of-range 0 1)
  replace-match("1. " nil nil nil 1)
  #f(compiled-function (struct old-struct item) #<bytecode 0x497071>)(((1 0 "1. " nil nil nil 6) (6 0 "2. " nil nil nil 13) (16 0 "1. " nil nil nil 20)) ((1 0 "1. " nil nil nil 6) (6 0 "1. " nil nil nil 13) (16 0 "2. " nil nil nil 20)) 16)
  org-list-struct-apply-struct(((1 0 "1. " nil nil nil 6) (6 0 "2. " nil nil nil 13) (16 0 "1. " nil nil nil 20)) ((1 0 "1. " nil nil nil 6) (6 0 "1. " nil nil nil 13) (16 0 "2. " nil nil nil 20)))
  org-list-write-struct(((1 0 "1. " nil nil nil 6) (6 0 "2. " nil nil nil 13) (16 0 "1. " nil nil nil 20)) ((1) (6) (16)))
  org-insert-item()
  eval((org-insert-item) nil)
  eval-expression((org-insert-item) nil nil 127)
  funcall-interactively(eval-expression (org-insert-item) nil nil 127)
  call-interactively(eval-expression nil nil)
  command-execute(eval-expression)
```

#### No indentation might be working as expected

```
1. a
b
2. c
```
Result:
```
1. a
2. 
b
2. c
````
At least there is no error message or freezing.

### System info
```
Org mode version 9.3.7 (9.3.7-4-gba6ca7-elpaplus @ c:/Users/username/.emacs.d/elpa/org-plus-contrib-20200622/)
GNU Emacs 26.3 (build 1, x86_64-w64-mingw32) of 2019-08-29
Windows 1903
```

**Comment by bzg** [comment]
Hi, thanks for pinging me. If this issue is directly related to org-mode, and if it hasn't been done already, can you post it to the mailing list (ie send an email to `emacs-orgmode@gnu.org`).  The mailing list is our central place for dealing with bugs. Thanks!

**Comment by duianto** [comment]
@bzg A message has been sent with the reproduction steps to that email address.

**Comment by bzg** [comment]
Great, thanks.

**Comment by duianto** [comment]
A fix has been applied upstream
https://code.orgmode.org/bzg/org-mode/commit/07a4a728628776202d58a48436da9f710e11cb8e
>list: Fix regression when inserting items
>* lisp/org-list.el (org-list-insert-item): Fix splitting item.
>* testing/lisp/test-org-list.el (test-org-list/insert-item): Add test.


I applied the fox manually and there's are no more freezing or error messages.

@dwuggh When the fix is part of the package update, if you can confirm that it works, then you can close this issue.

**Comment by duianto** [comment]
The fix is included in org 9.4. Update the packages to get it.

**Comment by bzg** [comment]
Indeed, thanks for the update.
